### PR TITLE
add user determined fees to peg-outs

### LINKED ISSUE #35: Make client architecture crash-safe
Make sure certain data is written to the DB *before* sending transactions to the federation. Otherwise funds can be lost if a crash happens.

### PR DESCRIPTION
**Pull request by jkitman**
Stops the fed from crashing due to peg-outs (fixes step 1 in #233)
- Peg-outs now only can happen once per epoch and only one peg-out is allowed per UTXO.
- Broadcasts are immediate after a peg-out is signed.
- Fees are exact and determined by the estimated min. sat/byte required by `FeeConsensus` (target confirmation time = 10 blocks) and the calculated transaction weight (no more profits/losses).
- Rather than always returning 404 if a tx failed the fed will store and return the `TransactionStatus::Rejected`, allowing clients react to tx errors.
- Stores pending coins and will reissue them in `reissue_pending_coins` if a tx is rejected (fixes #35)

### CODE REVIEW

**Review by NicolaLS — APPROVED** [review]
nice, I think the `fetch_peg_out_fees` naming is a bit confusing tough

File: client/client-lib/src/api.rs [hunk 1]
```diff
@@ -40,6 +42,13 @@ pub trait FederationApi: Send + Sync {
 
     /// Fetch the current consensus block height (trailing actual block height)
     async fn fetch_consensus_block_height(&self) -> Result<u64>;
+
+    /// Fetch the expected peg-out fees given a peg-out tx
+    async fn fetch_peg_out_fees(
+        &self,
+        address: Address,
+        amount: Amount,
```
[the remark this answers is not in the document]
Comment (jkitman) [reply client/client-lib/src/api.rs:50]: Good point!

File: client/client-lib/src/transaction.rs [hunk 2]
```diff
@@ -52,6 +95,16 @@ impl TransactionBuilder {
...
+        // move coins to pending state, awaiting a transaction
+        batch.append_from_iter(self.input_coins.iter().map(|(amount, coin)| {
+            BatchItem::delete(CoinKey {
+                amount,
+                nonce: coin.coin.0.clone(),
+            })
+        }));
+        batch.append_insert(PendingCoinsKey(self.tx.tx_hash()), self.input_coins);
+        batch.commit();
+
```
[the remark this answers is not in the document]
Comment (jkitman) [reply client/client-lib/src/transaction.rs:107]: Yeah you are right that could make sense because whenever we create coins (either as inputs or outputs) we want to ensure they are tied to a transaction so they don't get lost, that's why I moved coin input creation into the `TransactionBuilder` to enforce that through code.

File: minimint/src/consensus/conflictfilter.rs [hunk 3]
```diff
@@ -88,10 +86,30 @@ where
...
-                    return None;
+                    return Err(tx.clone());
+                }
+            }
+            if let Output::Wallet(_) = output {
+                if self.pegged_out {
+                    return Err(tx.clone());
                 }
+                self.pegged_out = true;
+            }
```
Comment (elsirion) [inline minimint/src/consensus/conflictfilter.rs:97]: These two if could probably be a match statement?

File: minimint-core/src/transaction.rs [hunk 4]
```diff
@@ -63,7 +63,7 @@ impl TransactionItem for Output {
     fn amount(&self) -> Amount {
         match self {
             Output::Mint(coins) => coins.amount(),
-            Output::Wallet(peg_out) => peg_out.amount.into(),
+            Output::Wallet(peg_out) => (peg_out.amount + peg_out.fees.amount()).into(),
```
Comment (elsirion) [inline minimint-core/src/transaction.rs:66]: Unrelated to this PR, but I feel like we should get rid of this API and only rely on what the modules return from `apply_*`

**Review by elsirion — COMMENTED** [review]
Great PR! Just a few questions :)

File: client/client-lib/src/mint/db.rs [hunk 5]
```diff
@@ -28,6 +30,24 @@ impl DatabaseKeyPrefixConst for CoinKeyPrefix {
     type Value = SpendableCoin;
 }
 
+#[derive(Debug, Clone, Encodable, Decodable)]
+pub struct PendingCoinsKey(pub TransactionId);
```
[the remark this answers is not in the document]
Comment (jkitman) [reply client/client-lib/src/mint/db.rs:34]: I think it makes sense to store all the input coins using a single tx id and then hit the `tx_status` api to determine whether the tx failed or not.  If anything maybe they should be called "in points"?

File: modules/minimint-wallet/src/lib.rs [hunk 6]
```diff
@@ -908,93 +883,83 @@ impl Wallet {
...
-            })
-            .collect::<Vec<_>>();
+                _ => return None, // Not enough UTXOs
+            }
+        }
 
-        // We might have selected too much value on the input side, so we need to pay the remainder
-        // back to ourselves.
-        let fees = feerate.calculate_fee(total_weight);
+        // We always pay ourselves change back to ensure that we don't lose anything due to dust
```
[the remark this answers is not in the document]
Comment (jkitman) [reply modules/minimint-wallet/src/lib.rs:939]: I think this should ensure we select more than the dust value https://github.com/fedimint/minimint/blob/c4097211b01329cb49fcd6facc6c6ea03952a702/modules/minimint-wallet/src/lib.rs#L927

File: client/client-lib/src/lib.rs [hunk 7]
```diff
@@ -259,31 +262,42 @@ impl<T: AsRef<ClientConfig>> Client<T> {
...
-
-        tx.input(&mut coin_keys, Input::Mint(coin_input));
-        tx.output(Output::Wallet(pegout_output));
+        let funding_amount = self.context.config.as_ref().fee_consensus.fee_peg_out_abs
+            + (peg_out.amount + peg_out.fees.amount()).into();
+        let coins = self.mint_client().select_coins(funding_amount)?;
+        tx.input_coins(coins, &self.context.secp)?;
+        tx.output(Output::Wallet(peg_out));
 
         self.submit_tx_with_change(tx, batch, &mut rng).await
```
[the remark this answers is not in the document]
Comment (jkitman) [reply client/client-lib/src/lib.rs:302]: I'm not sure whether it's best to ensure that all calls to `submit_transaction` occur before `apply_batch` writes to the DB on the client side OR whether we should store that as a `RejectedTransaction` in the fed's db as well.

File: minimint-core/src/transaction.rs (hunk 4 above)
Comment (jkitman) [reply minimint-core/src/transaction.rs:66]: Guess you are thinking modularly now :)

File: client/client-lib/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jkitman) [reply client/client-lib/src/lib.rs:302]: Currently `submit_tx_with_change` ensures this BTW, but other client code does not.

File: modules/minimint-wallet/src/lib.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (elsirion) [reply modules/minimint-wallet/src/lib.rs:939]: Right, I got confused. This ensures we'll need at least `dust_value` in change.

File: client/client-lib/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (elsirion) [reply client/client-lib/src/lib.rs:302]: > ensure that all calls to `submit_transaction` occur before `apply_batch` writes to the DB on the client side

I think that's the way to go. Do we always use `submit_tx_with_change`? Maybe that should be the only publicly available way to create mint inputs then (if it isn't already). 

> OR whether we should store that as a `RejectedTransaction` in the fed's db as well.

That would be too obvious of a DoS risk imo (I know we got others like offers, but still …).

File: minimint-core/src/transaction.rs (hunk 4 above)
Comment (elsirion) [reply minimint-core/src/transaction.rs:66]: After a day of haskell-like rust definitely :laughing:

File: client/client-lib/src/mint/db.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (elsirion) [reply client/client-lib/src/mint/db.rs:34]: I was just thinking we should make sure to not forget to include any coins in the DB entry, but I guess we shouldn't allow users to mess with coins anyway and just provide the "fund transaction, add change if needed and submit" functionality publicly and nothing else.

File: client/client-lib/src/lib.rs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jkitman) [reply client/client-lib/src/lib.rs:292]: > I think that's the way to go. Do we always use submit_tx_with_change? Maybe that should be the only publicly available way to create mint inputs then (if it isn't already).

With my latest change that should be the casea

File: client/client-lib/src/mint/mod.rs [hunk 8]
```diff
@@ -70,121 +71,60 @@ impl<'c> MintClient<'c> {
...
+        let final_tx = tx.build(
+            change_required,
+            batch.transaction(),
+            self.context.secp,
+            &self.context.config.tbs_pks,
+            rng,
+        );
+        let txid = final_tx.tx_hash();
+        let mint_tx_id = self.context.api.submit_transaction(final_tx).await?;
+        // TODO: make check part of submit_transaction
```
Comment (elsirion) [inline client/client-lib/src/mint/mod.rs:100]: I think we can remove that TODO now :smile:

File: client/client-lib/src/mint/mod.rs [hunk 9]
```diff
@@ -70,121 +71,60 @@ impl<'c> MintClient<'c> {
...
+        assert_eq!(
+            txid, mint_tx_id,
+            "Federation is faulty, returned wrong tx id."
+        );
 
-        tx.build(self.context.secp, &mut rng)
+        self.context.db.apply_batch(batch).expect("DB error");
```
Comment (elsirion) [inline client/client-lib/src/mint/mod.rs:106]: From a crash-safety pov we should first save and then submit imo. But that's a bigger issue (#35) and could be a separate PR, just something we need to keep in mind.

**Review by elsirion — APPROVED** [review]
Only reviewed the diff since my last review, only two nits. Waiting for @justinmoon to take a look.

**Comment by justinmoon** [comment]
LGTM. I sent kitman a couple messages on discord. In general it would be nicer to have smaller commits (I make huge commits too, working on it!). But looks like a great improvement.

File: client/client-lib/src/mint/mod.rs (hunk 9 above)
Comment (elsirion) [reply client/client-lib/src/mint/mod.rs:105]: I think what we want is:

1. "Stage" tx in DB (this PR introduces this intermediate state)
2. Submit tx
3. If an error occurs either on submit or the tx isn't confirmed for some reason we can later rollback the staged tx in the client DB. If the tx succeeds we finalize the staged tx in the client DB.

I assumed this `apply_batch` invocation was for step 1, but maybe I'm mistaken?
